collect all antrea-agent pod logs before redeploying antrea - e2e tests

siddhant94 opened the pull request:
Fixes #1226
Add log collection for antrea-agent pods when antrea is redeployed. This PR would lead to following behavioral change -  
If the user provides a log directory which was used for a previous run, existing contents (subdirectories for each test case) will be overridden.
To give overview of dir tree, running tests for TestIPSecTunnelConnectivity & TestPodConnectivityDifferentNodes (e2e) with `--logs-export-on-success` would result in the following dir tree -
```
── TestIPSecTunnelConnectivity
│   ├── beforeRedploy.Sep22-09-31-52
│   │   ├── k8s-node-master-antrea-agent-vnmwb-describe
│   │   ├── k8s-node-master-antrea-agent-vnmwb-logs
│   │   ├── k8s-node-worker-1-antrea-agent-kr9v4-describe
│   │   ├── k8s-node-worker-1-antrea-agent-kr9v4-logs
│   │   ├── k8s-node-worker-1-antrea-controller-6fbdccbd67-m8vht-describe
│   │   └── k8s-node-worker-1-antrea-controller-6fbdccbd67-m8vht-logs
│   ├── beforeRedeployWithIPsec.Sep22-09-30-38
│   │   ├── k8s-node-master-antrea-agent-zslgr-describe
│   │   ├── k8s-node-master-antrea-agent-zslgr-logs
│   │   ├── k8s-node-worker-1-antrea-agent-wtbps-describe
│   │   ├── k8s-node-worker-1-antrea-agent-wtbps-logs
│   │   ├── k8s-node-worker-1-antrea-controller-6df878958d-dj67t-describe
│   │   └── k8s-node-worker-1-antrea-controller-6df878958d-dj67t-logs
│   └── beforeTeardown.Sep22-09-32-48
│       ├── k8s-node-master-antrea-agent-9z6sx-describe
│       ├── k8s-node-master-antrea-agent-9z6sx-logs
│       ├── k8s-node-master-kubelet
│       ├── k8s-node-worker-1-antrea-agent-qm2d6-describe
│       ├── k8s-node-worker-1-antrea-agent-qm2d6-logs
│       ├── k8s-node-worker-1-antrea-controller-6df878958d-542ss-describe
│       ├── k8s-node-worker-1-antrea-controller-6df878958d-542ss-logs
│       └── k8s-node-worker-1-kubelet
└── TestPodConnectivityDifferentNodes
    └── beforeTeardown.Sep22-09-54-28
        ├── k8s-node-master-antrea-agent-9z6sx-describe
        ├── k8s-node-master-antrea-agent-9z6sx-logs
        ├── k8s-node-master-kubelet
        ├── k8s-node-worker-1-antrea-agent-qm2d6-describe
        ├── k8s-node-worker-1-antrea-agent-qm2d6-logs
        ├── k8s-node-worker-1-antrea-controller-6df878958d-542ss-describe
        ├── k8s-node-worker-1-antrea-controller-6df878958d-542ss-logs
        └── k8s-node-worker-1-kubelet

```

antoninbas commented on test/e2e/fixtures.go:
```diff
@@ -109,15 +137,7 @@ func setupTestWithIPFIXCollector(tb testing.TB) (*TestData, error) {
-}
-
-func exportLogs(tb testing.TB, data *TestData) {
+func exportLogs(tb testing.TB, data *TestData, dirName string) {
```
for this parameter, I would say that `logsSubDir` would be a better name than `dirName`

antoninbas commented on test/e2e/fixtures.go:
```diff
@@ -237,7 +257,11 @@ func exportLogs(tb testing.TB, data *TestData) {
-	exportLogs(tb, data)
+	exportLogs(tb, data, "beforeTeardown")
+	if empty, _ := IsDirEmpty(data.logsSubDir); empty {
+		tb.Logf("Removing empty Test name based sub-directory for logs '%s'", data.logsSubDir)
```
same thing as above, not a very useful or clear log message IMO

antoninbas reviewed (commented):
Thanks for working on this. Additional comments below:

In the PR description, please provide an example of the resulting log directory tree

In the commit message & PR description, please reference the corresponding issue with `Fixes #<NUM>` so that the issue gets closed automatically when we merge this (it also helps to know which specific issue is being addressed by this PR).

I think this README (https://github.com/vmware-tanzu/antrea/blob/master/test/e2e/README.md#running-the-tests) should be updated in this PR, as we have the following new behavior: if the user provides a log directory which was used for a previous run, existing contents (subdirectories for each test case) will be overridden.

siddhant94 commented on test/e2e/fixtures.go:
```diff
@@ -62,8 +63,35 @@ func ensureAntreaRunning(tb testing.TB, data *TestData) error {
+		tb.Errorf("Error creating logs directory '%s': %v", path, err)
+		return nil, err
+	}
+	data.logsSubDir = path
```
I think I have missed adding the setting up of log directories in `setupTestWithIPFIXCollector()`. This is also where this behaviour should be present, right?
And thank you for review comments, addressing these.

jianjuns commented:
Question - what you guys think about mount a separate host dir to /var/log/antrea and /var/log/openvswitch and collect logs there? @antoninbas @siddhant94

antoninbas commented:
@siddhant94 your PR is not rebased properly, as a result it shows a huge diff

antoninbas commented:
> Question - what you guys think about mount a separate host dir to /var/log/antrea and /var/log/openvswitch and collect logs there? @antoninbas @siddhant94

I have no strong preference. We don't need to mount anything though, we could just scp the contents from the nodes, since we already have ssh access. I would say it's better than the current approach if it gives us a way to cleanly separate the logs across test cases, which may require some work. The antrea-ovs container logs sometimes have useful information that we do not write to the host's filesystem.

siddhant94 commented:
> @siddhant94 your PR is not rebased properly, as a result it shows a huge diff

Fixed it and squashed all the commits

siddhant94 commented:
> Question - what you guys think about mount a separate host dir to /var/log/antrea and /var/log/openvswitch and collect logs there? @antoninbas @siddhant94

I literally have very little insight to provide a preference here. :)
Would definitely go through the logs (ovs) part to gain some meaningful insight as I get the chance.
Do let me know if the separation (antrea/ovs) needs to be implemented in this PR, would dig into further if that's the case.

jianjuns commented:
> Would definitely go through the logs (ovs) part to gain some meaningful insight as I get the chance.
> Do let me know if the separation (antrea/ovs) needs to be implemented in this PR, would dig into further if that's the case.

I am fine to merge your current implementation first, as it is complete for collecting controller and agent logs.
We can have a separate PR to cover OVS and strongSwan logs. And at that time we can re-evaluate whether to change to collecting logs from the host directory.

siddhant94 commented:
> > Would definitely go through the logs (ovs) part to gain some meaningful insight as I get the chance.
> > Do let me know if the separation (antrea/ovs) needs to be implemented in this PR, would dig into further if that's the case.
> 
> I am fine to merge your current implementation first, as it is complete for collecting controller and agent logs.
> We can have a separate PR to cover OVS and strongSwan logs. And at that time we can re-evaluate whether to change to collecting logs from the host directory.

siddhant94 commented:
@antoninbas @jianjuns I fixed the merged conflict as https://github.com/vmware-tanzu/antrea/pull/1160 got merged, but for some reason the Kind tests are failing with `FAIL: TestNetworkPolicyStats`, but the same tests run successfully when running as part of vagrant set-up locally. What is it that I am missing, any pointers would be very much helpful.

antoninbas commented:
@siddhant94 I have observed this test case failing as well. It was recently added. I have opened an issue for this: https://github.com/vmware-tanzu/antrea/issues/1309. Hoping @tnqn can take a look.

tnqn commented:
> @siddhant94 I have observed this test case failing as well. It was recently added. I have opened an issue for this: #1309. Hoping @tnqn can take a look.

The test became flaky because a following PR changed the behavior of termination of antrea pods, created #1312 to fix it.

antoninbas reviewed (dismissed):
LGTM

antoninbas commented:
@siddhant94 this PR is still tagged as a draft. Once Quan's PR (#1312) is merged, please rebase your PR and mark it as "Ready for review".

siddhant94 commented:
> @siddhant94 this PR is still tagged as a draft. Once Quan's PR (#1312) is merged, please rebase your PR and mark it as "Ready for review".

Done @antoninbas , rebased with master and removed draft tag. Thanks for review and corrections, it helped a  ton.

antoninbas commented:
/test-all

tnqn commented on test/e2e/connectivity_test.go:
```diff
@@ -192,7 +192,12 @@ func TestPodConnectivityDifferentNodes(t *testing.T) {
+	if enableIPSec {
+		exportLogs(t, data, "beforeEnableIPSec", false)
+	} else {
+		exportLogs(t, data, "beforeDisableIPSec", false)
```
redeployAntrea is used by `TestPodConnectivityAfterAntreaRestart` as well and might be used for other tests later. It might be a little confusing that other tests has a log directory named "beforeDisableIPSec" while it never enables it, how about "beforeRedeployAntreaWithIPSecEnabled" or "preRedeployAntreaWithIPSecEnabled"?

antoninbas replied:
+1, that's a good point

maybe just "beforeRedploy" and "beforeRedeployWithIPsecEnabled"

siddhant94 replied:
@tnqn @antoninbas Sounds good, i'll update it.

siddhant94 replied:
updated

siddhant94 commented on test/e2e/fixtures.go:
```diff
@@ -68,7 +68,31 @@ func ensureAntreaRunning(tb testing.TB, data *TestData) error {
+	if err := os.Mkdir(path, 0700); err != nil {
+		return err
+	}
+	return nil
```
updated :+1:

antoninbas commented on test/e2e/connectivity_test.go:
```diff
@@ -192,7 +192,12 @@ func TestPodConnectivityDifferentNodes(t *testing.T) {
+	if enableIPSec {
+		exportLogs(t, data, "beforeRedeployAntreaWithIPSecEnabled", false)
+	} else {
+		exportLogs(t, data, "beforeRedploy", false)
```
nit: I feel like for symmetry this should be "beforeRedployAntrea" or "Antrea" should be removed from "beforeRedeployAntreaWithIPSecEnabled". I recommend the latter to keep the directory name shorter. "beforeRedeployWithIPsec" would be best IMO.

siddhant94 replied:
Yes actually it's shorter and symmetric, thanks. Updated the PR.

antoninbas reviewed (approved):
LGTM. I think all of Quan's comments have been addressed as well.

antoninbas commented:
/test-all
